UniqueViolation on `ix_karton_analysis_object`

**Environment information**
* MWDB version (from `/about`):
v2.6.1
* Installation method:
  - [ ] mwdb.cert.pl service
  - [ ] From PyPi (`pip install mwdb-core`)
  - [x] From docker-compose
  - [ ] Other (please explain)
* Plugins installed:
None

**Behaviour the bug (what happened?)**

Summary: Tasks from `karton.mwdb-reporter` crash with `psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "ix_karton_analysis_object"` when there are multiple mwdb backends running and multiple mwdb-reporter replicas.

Infra setup: docker-compose with multiple `mwdb-core` instances and a single nginx reverse proxy. The nginx is a slightly modified version of the provided `mwdb-dev`, that round-robins the available backends; A karton pipeline with at least 2 `karton-mwdb-reporter` replicas.

Description: When the same object is queued multiple times (e.g. same config is extracted from same binary), and the tasks are consumed concurrently, there's a possibility of running into a `UniqueViolation`, an example of the trace of such error:

```
[ERROR] uWSGIWorker3Core0 - service.error_router:76 - Unhandled exception occurred
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1278, in _execute_context
    cursor, statement, parameters, context
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "ix_karton_analysis_object"
DETAIL:  Key (analysis_id, object_id)=(c6ebc495-fe32-47a8-8f53-b2ecff69de8a, 542646) already exists.


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1516, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1502, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/usr/local/lib/python3.6/site-packages/flask_restful/__init__.py", line 467, in wrapper
    resp = resource(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/flask/views.py", line 84, in view
    return current_app.ensure_sync(self.dispatch_request)(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/flask_restful/__init__.py", line 582, in dispatch_request
    resp = meth(*args, **kwargs)
  File "./mwdb/resources/__init__.py", line 47, in endpoint
    return f(*args, **kwargs)
  File "./mwdb/resources/__init__.py", line 31, in endpoint
    return f(*args, **kwargs)
  File "./mwdb/resources/config.py", line 281, in post
    return self.create_object(obj)
  File "./mwdb/resources/object.py", line 119, in create_object
    params, parent_object, share_with, attributes, analysis_id, tags
  File "./mwdb/resources/config.py", line 146, in _create_object
    tags=tags,
  File "./mwdb/model/config.py", line 56, in get_or_create
    tags=tags,
  File "./mwdb/model/object.py", line 570, in _get_or_create
    commit=False,
  File "./mwdb/model/object.py", line 453, in give_access
    obj.id, group_id, reason_type, related_object, related_user
  File "./mwdb/model/object.py", line 158, in create
    ObjectPermission.group_id == group_id,
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 3491, in scalar
    ret = self.one()
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 3458, in one
    ret = self.one_or_none()
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 3427, in one_or_none
    ret = list(self)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 3502, in __iter__
    self.session._autoflush()
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 1628, in _autoflush
    util.raise_(e, with_traceback=sys.exc_info()[2])
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 178, in raise_
    raise exception
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 1617, in _autoflush
    self.flush()
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2523, in flush
    self._flush(objects)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2664, in _flush
    transaction.rollback(_capture_exception=True)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/langhelpers.py", line 69, in __exit__
    exc_value, with_traceback=exc_tb,
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 178, in raise_
    raise exception
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2624, in _flush
    flush_context.execute()
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/unitofwork.py", line 422, in execute
    rec.execute(self)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/unitofwork.py", line 540, in execute
    self.dependency_processor.process_saves(uow, states)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/dependency.py", line 1177, in process_saves
    uowcommit, secondary_insert, secondary_update, secondary_delete
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/dependency.py", line 1239, in _run_crud
    connection.execute(statement, secondary_insert)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1014, in execute
    return meth(self, multiparams, params)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/sql/elements.py", line 298, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1133, in _execute_clauseelement
    distilled_params,
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1318, in _execute_context
    e, statement, parameters, cursor, context
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1512, in _handle_dbapi_exception
    sqlalchemy_exception, with_traceback=exc_info[2], from_=e
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 178, in raise_
    raise exception
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1278, in _execute_context
    cursor, statement, parameters, context
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely)
(psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "ix_karton_analysis_object"
DETAIL:  Key (analysis_id, object_id)=(c6ebc495-fe32-47a8-8f53-b2ecff69de8a, 542646) already exists.

[SQL: INSERT INTO karton_object (analysis_id, object_id) VALUES (%(analysis_id)s, %(object_id)s)]
[parameters: {'analysis_id': UUID('c6ebc495-fe32-47a8-8f53-b2ecff69de8a'), 'object_id': 542646}]
(Background on this error at: http://sqlalche.me/e/13/gkpj)
```

Worth noting that although the task crashes, the object was already inserted into the analysis by a previous identical task, (I think) there's no real consequence other than having a crashed task.

**Expected behaviour**

If the specific object is already connected to the karton analysis in MWDB, it should update any extra information (e.g. tags, relations) without crashing.

**Reproduction Steps**

Queue an object that generates 2 identical tasks e.g.: produces 2 identical configs, produces 2 identical binaries


EDIT: Made a clarification of how specifically this happens.

--- a/mwdb/model/object.py
+++ b/mwdb/model/object.py
@@ -924,10 +924,21 @@ class Object(db.Model):
         Assigns KartonAnalysis to the object
         """
         analysis, is_new = KartonAnalysis.get_or_create(analysis_id, self)
+
         if not is_new and analysis.id not in [
             existing.id for existing in self.analyses
         ]:
-            self.analyses.append(analysis)
+            db.session.begin_nested()
+            try:
+                self.analyses.append(analysis)
+                db.session.commit()
+                is_new = True
+            except IntegrityError:
+                # The same relationship was added concurrently
+                db.session.rollback()
+                db.session.refresh(self)
+                if analysis.id not in [existing.id for existing in self.analyses]:
+                    raise
 
         if commit:
             db.session.commit()

